### PR TITLE
Add fortran support

### PR DESCRIPTION
**Pull request by rubyzhou2014**
Add support for gcc, intel, PGI fortran compilers

### Fix or Enhancement?

Enhancement

- [x] New tests added
- [x] All tests passed

### Environment
- OS:
- Bear version: master

### CODE REVIEW

File: bear/main.py.in [hunk 1]
```diff
@@ -589,7 +625,10 @@ class Compilation:
 
         if command:  # not empty list will allow to index '0' and '1:'
             executable = os.path.basename(command[0])  # type: str
```
Comment (rizsotto) [inline bear/main.py.in]: I think this change is not intended. This was hint the type checker about the `parameters` is a list of strings. If you move this into another line, it won't mean the same thing.

File: bear/main.py.in [hunk 2]
```diff
@@ -470,10 +493,22 @@ def create_intercept_parser():
         default=[os.getenv('CXX', 'c++')],
         help="""Hint '%(prog)s' to classify the given program name as C++
         compiler.""")
+    
+    #add_argument for Fortan
+    parser.add_argument(
+        '--use-f',
+        metavar='<path>',
+        dest='use_f',
+        action='append',
```
Comment (rizsotto) [inline bear/main.py.in]: The current default `fXX` does not look good. I don't think there is such compiler... Why not just using `f95`?

File: bear/main.py.in [hunk 3]
```diff
@@ -115,9 +117,18 @@ COMPILER_PATTERNS_CXX = (
...
+COMPILER_PATTERNS_F = (
+    re.compile(r'^(f95)$'),
+    re.compile(r'^(gfortran)$'),
+    re.compile(r'^(ifort)$'),
+    re.compile(r'^(pg)(f77|f90|f95|fortran)$')
+    )
+
 TRACE_FILE_PREFIX = 'execution.'  # same as in ear.c
 
-C_LANG, CPLUSPLUS_LANG, OTHER = range(3)
```
Comment (rizsotto) [inline bear/main.py.in]: Could you write `FORTRAN_LANG` instead of `F` please? It's more verbose, but helps to understand.

File: bear/main.py.in [hunk 4]
```diff
@@ -115,9 +117,18 @@ COMPILER_PATTERNS_CXX = (
     re.compile(r'^(g|)xl(C|c\+\+)$'),
 )
```
Comment (rizsotto) [inline bear/main.py.in]: Could you write `COMPILER_PATTERNS_FORTRAN`?

File: bear/main.py.in [hunk 5]
```diff
@@ -92,11 +92,13 @@ IGNORED_FLAGS = {
 
 }  # type: Dict[str, int]
 
+
 # Known C/C++ compiler wrapper name patterns.
 COMPILER_PATTERN_WRAPPER = re.compile(r'^(distcc|ccache)$')
 
 # Known MPI compiler wrapper name patterns.
-COMPILER_PATTERNS_MPI_WRAPPER = re.compile(r'^mpi(cc|cxx|CC|c\+\+)$')
+
```
Comment (rizsotto) [inline bear/main.py.in]: Is this change required? Can you provide me links for MPI compiler wrappers called `mpifort` or `mpif77`?

**Comment by chunhualiao** [comment]
@rizsotto @rubyzhou2014 is working with me to add this feature. Could you please provide an example integration test so she can follow? I guess she can use the NPB fortran test suites for the test.

File: bear/main.py.in (hunk 5 above)
Comment (chunhualiao) [reply bear/main.py.in]: @rizsotto I googled mpifort and found https://www.mpich.org/static/docs/v3.1.2/www1/mpifort.html . You should be able to find mpif77 the same way.

**Comment by rubyzhou2014** [comment]
> Thank you for making this PR!
> 
> * Could you please look at the CI output, there are some formatting problems.
> * You've made empty lines before many of your changes. Could you remove those? That would help to see what are the real changes.
> * Please write `fortran` and not just `f` for variable names and function names.
> * Please also try to write integration test (Will require to install fortran compiler into the CI env too.)

@rizsotto  Thanks for your reply. I am trying to fix the formatting/empty lines/naming problems and will send another pull request about it later. 
I am currently working with @chunhualiao as an intern. He sent me issue #241 as my side project. I plan to complete the CI test in the following two weeks. By the way, as Chunhua mentioned, can you show me some CI test sample that I can follow? Or is there any specific things that I need to pay attention to? Thanks a lot.

**Comment by rizsotto** [comment]
Hey @rubyzhou2014 , thanks for the update. I'm away from keyboard. Will come back to you next week.

File: bear/main.py.in (hunk 1 above)
Comment (rizsotto) [reply bear/main.py.in]: It really has to be like this.
```
parameters = command[1:]  # type: List[str]
```
Could you please revert this line change?

**Comment by rizsotto** [comment]
Thanks for the update, it really looks good.

About the integration tests: they are live under the `tests` directory. (Those would be the examples I can offer.) It is written on the [lit](https://llvm.org/docs/CommandGuide/lit.html) tool. What I would like to ask from you is: to implement a small Fortran project, with GNU makefile. Then create a test which runs Bear and expects the compilation database content as you specify.

**Comment by rubyzhou2014** [comment]
> Thanks for the update, it really looks good.
> 
> About the integration tests: they are live under the `tests` directory. (Those would be the examples I can offer.) It is written on the [lit](https://llvm.org/docs/CommandGuide/lit.html) tool. What I would like to ask from you is: to implement a small Fortran project, with GNU makefile. Then create a test which runs Bear and expects the compilation database content as you specify... Once it runs on your machine, you will need to update the travis config file to install the tools that the test requires.

Thanks for your info. I will check it out. BTW, I just solved the formatting problem above. Thanks.

**Comment by rubyzhou2014** [comment]
> Can you tell me, what are the binary files?

What do you mean by binary? 
By the way, I just modified the test case and submitted a pull request.  Thanks.

**Comment by rizsotto** [comment]
Thanks @rubyzhou2014 ! Will look into it this week.

> > Can you tell me, what are the binary files?
> 
> What do you mean by binary?

The PullRequest [shows](https://github.com/rizsotto/Bear/pull/253/files) 3 files which GitHub claims as binary files.

**Comment by rubyzhou2014** [comment]
> Thanks @rubyzhou2014 ! Will look into it this week.
> 
> > > Can you tell me, what are the binary files?
> > 
> > 
> > What do you mean by binary?
> 
> The PullRequest [shows](https://github.com/rizsotto/Bear/pull/253/files) 3 files which GitHub claims as binary files.

They are generated automatically by Mac ios(https://hints.binaryage.com/how-to-remove-ds-store-files-from-a-git-repo/). I've removed these DS_store binary files and add "DS_store" into the file .gitignore.

**Comment by rizsotto** [comment]
thanks to clear things up. I've still have one file in the commit (named `@`) which seems to be a not needed file.

**Comment by rizsotto** [comment]
Planing to release this on the weekend (or next week). It's now on the `devel` branch. If you can test and give me some feedback, I would really appreciate it.

- Squashed all your commits into one.
- Was made small changes on top of yours.

Thanks to contribute to this project. Since I don't have any experience on this field, I could not make it by myself. Hope you enjoyed to work on the code. Please give me some feedback if you have time on the code, the environment and my review activity.

Thanks again!
